### DISCUSSION TITLE
How to use webhooks

### DISCUSSION THREAD

**Discussion by anislanguer** [fastapi/fastapi#12121] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request

app = FastAPI()

@app.webhooks.post("my-webhook")
async def my_webhook(request: Request):
    data = await request.json()
    # Process the incoming data
    return {"message": "Webhook received", "data": data}

if __name__ == "__main__":
    import uvicorn
    print(app.routes)
    uvicorn.run(app, host="0.0.0.0", port=9000)
```

### Description

Hello i wanted to use that code to generate a webhook but i didnt find how to call that webhook can someone help me ?

### Operating System

Windows

### FastAPI Version

0.112.2

### Pydantic Version

1.10.13

### Python Version

python 3.9.7

**Comment by YuriiMotov** [comment, marked as answer]
I created an example of how I understand webhooks should be implemented with FastAPI:

**Main app (sender of notifications):**
```
from typing import Annotated, Literal
from fastapi import Body, FastAPI, Request
import httpx
from pydantic import BaseModel

app = FastAPI()

fake_db = {}

async def set_webhook_url_for_user(user_id: int, webhook_url: str):
    fake_db[user_id] = webhook_url

async def get_webhook_urls():
    return list(fake_db.values())


class NewUserNotification(BaseModel):
    user_name: str

class WebHookResponse(BaseModel):
    status: Literal["ok", "error"]

# We create the route below just to document the format of webhook url we expect
# from the client
@app.webhooks.post("new-user-webhook")
async def my_webhook(notification_data: NewUserNotification):
    """
    When a new user is added we'll send you a POST request with this
    data to the URL that you register for the event `new-user-webhook` via `set-webhook`
    endpoint.
    """
    pass


@app.post("/set-webhook")
async def set_webhook(
    user_id: int,
    webhook_url: Annotated[
        str,
        Body(
            description="See info about webhook at 'Webhooks\\new-user-webhook' section",
            embed=True,
        )
    ],
):
    await set_webhook_url_for_user(user_id, webhook_url)
    return "Ok"


@app.post('/add-user')
async def create_user(user_name: str):
    # TODO: Check everyfing and add user to the DB..

    # Notify clients by sending notifications to webhook urls
    notification_data = {
        "notification_data": {"user_name": user_name}
    }
    webhook_urls = await get_webhook_urls()
    async with httpx.AsyncClient() as client:
        for webhook_url in webhook_urls:
            resp = await client.post(webhook_url, json=notification_data)
            if (resp.status_code != 200) or (resp.json().get("status") != "ok"):
                print(f"Failed to send notification to {webhook_url}")
```

**Client-side app (receiver of notifications):**

```
from fastapi import FastAPI, Request

app = FastAPI()

@app.post('/receive-notification')
async def receive_notification(request: Request):
    data = await request.json()
    # Process the incoming data
    print(f"Notification has been received: {data}")
    return {"status": "ok"}

```

**To test it:**
1. run both apps on different ports (e.g. 8000 for main app and 8001 for client-side app).
2. go to [127.0.0.1:8000#docs](http://127.0.0.1:8000/docs#) and send request to `set-webhook` endpoint. Specify `user_id=1` and body data `{"webhook_url": "http://127.0.0.1:8001/receive-notification"}`.
3. then send request to `add-user`. Specify `user_name=somebody`.
4. take a look at the console of client-side app. You should find `Notification has been received: {'notification_data': {'user_name': 'somebody'}}` there

**Reply by anislanguer** [reply]
then what is exactly the untility of the webhook here ? 
one can directly execute the receive-notification endpoint

**Reply by Abdi-Arslene** [reply]
@YuriiMotov  So let's say i have a platform that needs a webhook url, i just need to set up an endpoint that's all ?

**Reply by YuriiMotov** [reply]
> then what is exactly the untility of the webhook here ? one can directly execute the receive-notification endpoint

The thing is that you need to subscribe once  (call `set-webhook`) and you will be notified every time somebody create new user calling `add-user` (if we are talking about my code exanple)

**Reply by YuriiMotov** [reply]
> @YuriiMotov So let's say i have a platform that needs a webhook url, i just need to set up an endpoint that's all ?

If you are using a platform that allows to specify webhook url to notify you about any events, then you just need to create an endpoint as I did for client-side app.

If you are creating the platform that will notify others by sending notifications via webhook urls, you need 1) create mechanism to set client's webhook url, 2) perform a request to that stored webhook url every time you need to notify client. 3) optionally, you can add to openapi schema information on how client should implement webhook endpoint (http method, request parameters, response schema).

**Reply by RickardCardell** [reply]
Thanks for the example! I also thought that the Fastapi documentation is a bit vague of the purpose of declaring webhooks.

I'll add just a minor note about a benefit of using the webhook-declarations. The webhook above is documented to send notification_data of type `NewUserNotification` to the registered urls.  
If the code in the service actually uses that class as well, it ensures that the code also follows that contract.

I.e. if the specificication of `NewUserNotification` changes, e.g. another fields is added, it will automatically be reflected in the webhook
```
@app.post('/add-user')
async def create_user(user_name: str):
    # TODO: Check everyfing and add user to the DB..

    # Notify clients by sending notifications to webhook urls
    notification_data = NewUserNotification(user_name=user_name)  # <----------------------
    webhook_urls = await get_webhook_urls()
    async with httpx.AsyncClient() as client:
        for webhook_url in webhook_urls:
            resp = await client.post(webhook_url, json=notification_data.model_dump_json())
            if (resp.status_code != 200) or (resp.json().get("status") != "ok"):
                print(f"Failed to send notification to {webhook_url}")
```
